File: solutions/python/proverb/1/proverb.py
def not_none(value) -> bool:
    return value is not None


def proverb(*args, qualifier=None) -> str:
    main = [f"For want of a {name1} the {name2} was lost." for name1,
            name2 in zip(args, args[1:])]

    last_name = ' '.join(
        filter(not_none, [qualifier, args[0] if len(args) > 0 else None]))
    last = f"And all for the want of a {last_name}."

    return [*main, last] if len(args) > 0 else []

File: solutions/ruby/proverb/3/proverb.rb
class Proverb
  def initialize(*chain, qualifier:nil)
    @chain = chain
    @final = [qualifier, chain.at(0)].compact.join(' ')
  end

  def proverb(item1, item2)
    "For want of a #{item1} the #{item2} was lost."
  end

  def final(item)
    "And all for the want of a #{item}."
  end

  def to_s
    @chain.each_cons(2)
    .map{|item1, item2| proverb(item1, item2)}
    .push(final(@final))
    .join("\n")
  end
end